rust/README.md

```diff
@@ -0,0 +1,24 @@
+# substrait-rs
+
+Experimental Rust bindings for [substrait](https://substrait.io/).
+
+## Build & Test
+
+``` bash
+cargo test
+```
+
+## Generate Documentation
+
+``` bash
+cargo doc --no-deps
+```
+
+## Publishing the crate
+
+We need to specify `--allow-dirty` when publishing because we copy the `.proto` files into the local directory but
```

Comment:
Is there way we can do this without the copy? E.g. just give the relative path to the files (or symlink the files?)

Reply:
No, at least not that I know of.

Rust "crates" that get published to crates.io are essentially just tarballs with the file extension `.crate` that contains everything needed to compile from source, so we must include the `.proto` files in this crate tarball.

If we reference the proto files using a relative path that is outside of the `rust` directory then it will not make sense in the context of running against the extracted tarball.

One workaround would be to put the `Cargo.toml` in the root of the repository but that does not seem ideal.

If we had these bindings in a separate repo then we could use a git submodule to pull the main repo with the proto files. That is perhaps the cleanest option and we can use version control to make sure we are referencing a specific version of the proto files.

Reply:
I also tried with symlinks but it leads to the same issue where Cargo sees local files that are not checked in:

```
$ cargo publish --dry-run
    Updating crates.io index
error: 9 files in the working directory contain changes that were not yet committed into git:

build.rs
proto/substrait/function.proto
proto/substrait/capabilities.proto
proto/substrait/type.proto
proto/substrait/type_expressions.proto
proto/substrait/parameterized_types.proto
proto/substrait/algebra.proto
proto/substrait/plan.proto
proto/substrait/extensions/extensions.proto

to proceed despite this and include the uncommitted changes, pass the `--allow-dirty` flag
```

Reply:
I updated https://github.com/andygrove/substrait-rs to demonstrate the git submodule approach, which works well. I also added the GitHub action there.

Reply:
Got it, thanks

Reply:
@jacques-n So, what's the plan for this? Multiple repositories + submodules (presumably with all repositories moved to the substrait-io organization) or merging this PR?

I'm asking because I'm running into the same problems with the validator and its various language bindings now. My workarounds don't work 100% yet, and I've already had to resort to some *ugly* stuff.

Also, while trying to get things working right for the validator, I realized that the approach in this PR doesn't work 100% either:

 - `build.rs` is not run locally by `cargo package`/`cargo publish`, at least not until it tries to verify that the .crate file it built will compile out of context. For the crate to be packaged correctly, it's essential that `cargo build` is run first, or it will fail validation because the files don't exist at best (just run `cargo package` in a clean clone of the repo; it will fail), or publish with out-of-date .proto files that were copied earlier at worst.
 - I don't think you'd be able to use these bindings by specifying the git repository as the dependency, since `Cargo.toml` is not in the repository root. (I've already run into similar problems with other tools that assume that it lives there, so I actually already had to put a workspace `Cargo.toml` and the associated `Cargo.lock` file in the root. As more and more bindings will likely be added over time, I suspect this is going to create a lot of clutter.)
 - Highly unlikely, but if the `../proto` directory would ever happen to exist in a user's build environment when the .crate file is built as a dependency, `build.rs` might behave erratically. The published crate should in general never be probing outside of its build directory (except to look for system dependencies like protoc, I guess).

I don't think there *is* a good way to do this without resorting to submodules. The "best" solution I suppose would be to use `cargo xtask codegen` to do the copy rather than `build.rs`, but `cargo-xtask` is not a standard thing, and `cargo package` would still complain without `--allow-dirty`. `cargo` just doesn't like crates relying on context outside of the manifest directory, and with good reason IMO. Likewise for the Python bindings I have for the validator, and it probably applies to many other languages as well.

If we *do* go the multiple-repository route, I would suggest to move the proto and yaml files into a separate repository as well. That way, the main repository can have submodules for it *and* the language bindings (and validator), so a user can just get everything in one go if they want, and so we can have some centralized CI and release logic to help ensure everything stays in sync.

Reply:
> * I don't think you'd be able to use these bindings by specifying the git repository as the dependency, since `Cargo.toml` is not in the repository root.

That part actually does work. Cargo is smart enough to search the repo for the appropriate `Cargo.toml`. We are depending on this PR branch from [datafusion-substrait](https://github.com/datafusion-contrib/datafusion-substrait) for example. 

However, I agree with your other points and would also prefer so see these bindings in their own repo with a submodule to pull in the proto files.

Reply:
Since it doesn't look like there's going to be a resolution on this anytime soon, I had another look at the build logic for the validator. I ended up doing pretty much what you're doing now, also copying over the files in build.rs. I did figure out a way to resolve the third point I made, namely by adding a sort of marker file in the crate directory that's excluded in the package, in order to determine whether build.rs is being run from the git repository or from a downloaded crate file. If the file doesn't exist, it knows not to probe anything outside of its manifest directory. [Here](https://github.com/jvanstraten/substrait/blob/2e1c1172c3ff3d327ac255db56176989f6fa8deb/validator/rs/build.rs)'s my file for reference.

Good to know that Cargo itself doesn't require the manifest file to be at the root, by the way. It's smarter than I thought, I guess :) Unfortunately some of the third-party stuff I'm using isn't, I think, so I don't think I can easily move my workspace file to a nicer location in my soon-to-be PR.